### DISCUSSION TITLE
Hugmyndir að námskeiði og bættum glósur

### DISCUSSION THREAD

**Discussion by AldaAsl** [HI-IDN/IDN302G-GitBook#5] [Hugmyndir]
### Námsfyrirkomulag og verkefni

Ég myndi byrja önnina að kynna fyrir nemendum hvernig hægt er að nýta sér þennan áfanga á skemmtilegan hátt og sýna hvernig hann gagnast fólki á atvinnumarkaði (nefna eh fyrirtæki og hvernig þau notast við þessa hæfni/tækni en ekki bara tala um hversu gagnlegt þetta er og sýna eh gröf sem maður skilur ekkert í því það eykur ekki áhuga). 

Síðan myndi ég sýna uppá töflu í tíma hvernig maður getur unnið í einhverju verkefni t.d. í VSCode og biðja nemendur að gera slíkt hið sama í sinni tölvu. Getur þá sýnt hvernig maður pushar, hvað gerist ef annar hópmeðlimur pullar og getur sýnt ef þú settir verkefnið kannski á vitlaust branch, hvernig hægt væri að breyta því og einnig sýnt hvernig þú myndir breyta verkefninu sem þú værir búin að setja inná main. Á þennan hátt færðu alla nemendur til að fylgjast vel með og læra sjálfir þar sem þeir fylgja því sem þú gerir. 

Eftir þetta, myndi ég setja fram lítil verkefni í byrjun sem væru unnin í tímum og þá gætu allir hjálpast að. Mér fannst ég t.d. læra mest á því að gera einstaklings-regex verkefnið því þá unnum við verkefnið í tíma og gátum hjálpast að en einnig fengið aðstoð. 
Á þennan hátt fengi maður betra tækifæri til að fá tilfinningu fyrir github og nær að skilja hvernig það virkar (sleppa að láta okkur horfa á löngu myndböndin á yt).

Eftir nokkur svona lítil verkefni sem unnin eru í tímum, myndi ég hafa nokkur svona TBL verkefni sem mér fannst fjölbreytt og gagnleg en þó voru þau mörg hver of erfið og flókin en ég held að með þessum fyrstu verkefnum sem unnin eru í sameiningu og í tímum yrðu þau auðveldari.

Seinna myndi ég leyfa fólki að velja um frumleg og skemmtileg verkefni á þeirra áhugasviði líkt og að búa til vefsíðu eða tölvuleik með þeim kröfum sem gerðar eru til þessa námskeiðs. Þannig væru verkefnin meira unnin á þeirra forsendum og bæði áhugi og vilji nemenda til að skilja hvað námsefnið gengur útá verður mun meiri. 

Annars fannst mér verkefnin mjög fjölbreytt og sum skemmtileg (aðallega Capstone). Eftir á að hyggja lærði maður mjög margt. 

### Myndbönd
Væri einnig mjög gott að fá myndbönd í byrjun sem útskýra hvernig maður setur forritunin upp þar sem það tók einn dag að setja upp Github og annan dag að setja upp DataGrip. Einnig má sýna betur notkun Postgres þar sem maður var mjög lost þar í verkefninu tímataka. 

Einnig sýna hvernig hægt væri að takast á við helstu vandamál í Terminal með skipunum líkt og –force, reset og checkout. 

### Skipanir
Það væri mjög hjálpsamt að hafa kafla í glósunum inná Gitbook sem sýnir alls kyns nytsamlegar skipanir í Terminal ef eitthvað fer úrskeiðis á github sem á svo oft til með að gerast. Gætir haft myndir og útskýringar á þeim. 

**Ef pushað er vitlausu verkefni**
git reset --hard HEAD~1
git add <skrá>
git push

**Ef pushað verkefni á vitlaust branch**
	git reset --hard HEAD~1
git push origin <vitlaust-branch> --force
git checkout <rétt-branch>
git cherry-pick <commit-hash>
git push origin <rétt-branch>

**Ef gleymdist að bæta breytingum í commit**
 	git add <skrá>
git commit --amend --no-edit
git push origin <branch> --force

**Ef þú vilt fá eldri útgáfu af verkefninu**
git revert HEAD
git push origin <branch>
git reset --hard <commit-hash>
git push origin <branch> --force

**Ef pushað var óvart á main**
	git checkout -b <nýtt-branch>
git push origin <nýtt-branch>
git checkout main
git reset --hard origin/main
git push origin main --force

**Stofnað branch með röngu nafni**
	git branch -m <gamalt-nafn> <nýtt-nafn>
git push origin --delete <gamalt-nafn>
git push origin <nýtt-nafn>

Og útskýra vel að það er mikilvægt að forðast ofnotkun á --force, sérstaklega á sameiginlegum branchum. Gott að nota gitstatus og git log til að fá yfirsýn yfir stöðu og sögu breytinga. Einnig ef það kemur merge conflict, hvernig hægt er að vinna úr því. 
Væri líka æði ef hægt væri að redda 1-2 einstaklingum sem geta aðstoðað í einhverjum tímum þar sem oft þarf að bíða lengi eftir aðstoð í þessum áfanga. 

Annars var glósuheftið mjög hjálplegt, Regex101 verkefnið sniðugt og Capstone þar við gátum nýtt þekkingu okkar á efninu og skoðað efni við okkar hæfi.

**Comment by tungufoss** [comment]
Takk kærlega fyrir ábendingarnar @AldaAsl, þær eru bæði gagnlegar og vel ígrundaðar. Ég sé strax hvernig ég get tekið þær til mín og notað til að endurbæta námskeiðið.

Góður punktur með fyrsta tímann – ég mun skoða að byrja næsta námskeið með því að sýna frekar dæmi um hve fjölbreytt mælaborð sem þið gerðuð með Capstone verkefninu ykkar (með ykkar leyfi að sjálfsögðu). Það myndi gefa nemendum skýrari mynd af því hversu yfirgripsmikið efnið er og hvernig það tengist raunverulegum verkefnum, frekar en að byrja með almennri og frekar þurri glærukynningu á gagnagreiningu.

Ég viðurkenni að vinnuálagið hefur verið stundum orðið mikið, og ég mun skoða að byrja námskeiðið á smærri verkefnum sem unnin eru í tímum. Það gæti hjálpað til við að byggja upp skilning og sjálfstraust áður en flóknari TBL-verkefnin taka við.

Einstaklings regex verkefnið var í upphafi ekki hluti af námsáætlun, en var bætt inn þar sem TBL verkefnið var ekki að ganga jafnvel og ég var að vonast eftir. Það var skref í rétta átt, en ég sé klárlega að sambland af einstaklings- og hópverkefnum gæti verið enn betra og hentað nemendum betur.

Þetta er allt saman mjög gangvel endurgjöf sem ég tek til mín og mun nýta til að bæta námskeiðið enn frekar í næstu ítrun. 

PS. Ég mun bæta við þessum *git* skipunum í glósuheftið.

**Comment by tungufoss** [comment]
Takk fyrir ítarlega endurgjöf! Áþreifanlegu efnisóskirnar fóru í issue #55 svo hægt sé að vinna úr þeim. Loka þessari umræðu sem afgreiddri.
